### PR TITLE
pkg: Add and integrate BulkRowWriter processor into CTAS execution.

### PR DESCRIPTION
**Pull request by adityamaru27**
This change introduces a BulkRowWriter processor which uses
AddSSTable to write rows from a RowSource to the target
table. This is in an attempt to make the CREATE TABLE ... AS
statement more scalable.

### CODE REVIEW

File: pkg/sql/logictest/testdata/planner_test/show_trace [hunk 1]
```diff
@@ -134,7 +134,7 @@ statement ok
 SET tracing = on,kv,results; UPDATE t.kv2 SET v = v + 2; SET tracing = off
 
 query TT
-SELECT operation, regexp_replace(message, '(\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d\.)?\d\d\d\d\d+', '...PK...') as message
```
Comment (adityamaru27) [inline pkg/sql/logictest/testdata/planner_test/show_trace]: Using the row converter changes the way row_id's are generated for hidden PK's. They are generated using `GenerateUniqueID` which required this change in the logic test.

**Review by dt — COMMENTED** [review]
> Reviewed 10 of 10 files at r1, 7 of 7 files at r2, 7 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38374)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27, @dt, and @lucy-zhang)

___
*[pkg/sql/create_table.go, line 1652 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LinEm374iTp6ZatP2D3:-LinEm374iTp6ZatP2D4:bhley45) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c528870bdbc83caf174b904bcf17573a12f8bac/pkg/sql/create_table.go#L1652)):*
> ```Go
> }
> 
> func newBulkRowWriterProcessor(
> ```

for better or worse, most of the time we define processors in `distsqlrun` and only use the function-pointer injection route for the ones that have to be defined elsewhere due to hard rules about dependency edges, like OSS vs CCL. Is there anything motivating putting this in `sql` instead of `distsqlrun`? Anything it depends on in `sql`?

in either case, I'd pull it into own file.
___
*[pkg/sql/create_table.go, line 1744 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LinG8I69Zh1mVu5z5E-:-LinG8I7Cj76zex-Jc26:bh75xio) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c528870bdbc83caf174b904bcf17573a12f8bac/pkg/sql/create_table.go#L1744)):*
> ```Go
> 
> 		g = ctxgroup.WithContext(ctx)
> 		g.GoCtx(func(ctx context.Context) error {
> ```

is sp.out threadsafe?

2¢: I might find this easier to follow if we
a) pull the first half of this inline func into a private method on the processor which also inlines ingestKVs.
b) I'd probably have that method just set a BulkOpSummary field in the processor when it is done, instead of emitting directly.
c) I'd then serialize and emit the row part done in `Run`, after calling g.Wait (so you don't have to worry about a race).
___
*[pkg/sql/create_table.go, line 1778 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LinHgLNEElzvM0IMNFJ:-LinHgLNEElzvM0IMNFK:b-ja7gx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c528870bdbc83caf174b904bcf17573a12f8bac/pkg/sql/create_table.go#L1778)):*
> ```Go
> 		})
> 
> 		for {
> ```

right now if the ingesting goroutine gets an error, I think this will keep pushing rows at it and might deadlock on the channel. I think you might want to also run this loop in a task on the error group so that its context is hooked up to cancel on error return (and the other way too -- if this errors, we should teardown the ingesting goroutine too).
___
*[pkg/sql/distsqlrun/processors.go, line 1244 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LiFVZ5UAWbBVxQCZzQr:-LiFVZ5UAWbBVxQCZzQs:bcs0vqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d38628e6617e855bb05efd53dc7d79133640f19/pkg/sql/distsqlrun/processors.go#L1244)):*
> ```Go
> var NewCSVWriterProcessor func(*FlowCtx, int32, distsqlpb.CSVWriterSpec, RowSource, RowReceiver) (Processor, error)
> 
> // NewBulkRowWriterProcessor is externally implemented.
> ```

This is usually done to get around CCL vs OSS, where we inject the proc via this hook and an init func. This is a pure OSS proc, so I don't think we need that song and dance.

**Review by adityamaru27 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38374)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27, @dt, and @lucy-zhang)

___
*[pkg/sql/create_table.go, line 1744 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LinG8I69Zh1mVu5z5E-:-LisGkXI72CE87u5GWdq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c528870bdbc83caf174b904bcf17573a12f8bac/pkg/sql/create_table.go#L1744)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

is sp.out threadsafe?

2¢: I might find this easier to follow if we
a) pull the first half of this inline func into a private method on the processor which also inlines ingestKVs.
b) I'd probably have that method just set a BulkOpSummary field in the processor when it is done, instead of emitting directly.
c) I'd then serialize and emit the row part done in `Run`, after calling g.Wait (so you don't have to worry about a race).



</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1778 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LinHgLNEElzvM0IMNFJ:-Liq2BO-9WX6IqbC-S60:by8q00t) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c528870bdbc83caf174b904bcf17573a12f8bac/pkg/sql/create_table.go#L1778)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

right now if the ingesting goroutine gets an error, I think this will keep pushing rows at it and might deadlock on the channel. I think you might want to also run this loop in a task on the error group so that its context is hooked up to cancel on error return (and the other way too -- if this errors, we should teardown the ingesting goroutine too).
</blockquote></details>

thats a cool catch, good to know. Done.

**Review by thoszhang — COMMENTED** [review]
I'm not very familiar with this part of the code, but it LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-:-LitXPjw9vK-W6JvvjOH:b-gwx29q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27, @dt, and @lucy-zhang)

___
*[pkg/sql/bulk_row_writer.go, line 82 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LitTllDDp3hoch-p_wY:-LitTllDDp3hoch-p_wZ:b-si0sp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c50c53cd89c44fcfe4b2c2cdef6716b23a0a277/pkg/sql/bulk_row_writer.go#L82)):*
> ```Go
> 	// ingestKvs drains kvs from the channel until it closes, ingesting them using
> 	// the BulkAdder. It handles the required buffering/sorting/etc.
> 	ingestKvs := func() error {
> ```

Is there a difference between defining this function and then calling it immediately afterward, and just replacing it by the function definition?

**Comment by adityamaru27** [comment]
> Is there a difference between defining this function and then calling it immediately afterward, and just replacing it by the function definition?

I don't think so, I just did it to modularize the kv ingestion logic. However, I just realized that this pattern is used for places where we reuse a method several times. Should I pull out the definition?

**Review by dt — APPROVED** [review]
> Reviewed 1 of 11 files at r3, 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38374)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27)

___
*[pkg/sql/bulk_row_writer.go, line 70 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LjHVsCs6cgH4uHDhuKZ:-LjHVsCs6cgH4uHDhuK_:b-c86q4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c50c53cd89c44fcfe4b2c2cdef6716b23a0a277/pkg/sql/bulk_row_writer.go#L70)):*
> ```Go
> }
> 
> func (sp *bulkRowWriter) doRun(ctx context.Context, kvCh chan []roachpb.KeyValue) error {
> ```

nit: I'd call this method `ingestLoop` or something descriptive since it doesn't really _run_ the whole process.

That said, I think you're also on the right track going with the pattern of a `doRun` method that has an error return and can thus have early returns on error, since you can then just call it in `Run` and cleanup some of the named errors and conditionals deciding what to return.
___
*[pkg/sql/bulk_row_writer.go, line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LjHXATW-gBbp6EwKhwz:-LjHXATW-gBbp6EwKhx-:bj84st6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c50c53cd89c44fcfe4b2c2cdef6716b23a0a277/pkg/sql/bulk_row_writer.go#L118)):*
> ```Go
> 	var kvCh chan []roachpb.KeyValue
> 	var g ctxgroup.Group
> 	err := func() error {
> ```

This is a lot of inline anon function along with the added layer below for the GoCtx call.

In my experience, big inline funcs, particularly closing over each other, can make it tricky to see what vars closed over vs local, what is shadowed, etc. and can sometimes lead to subtle bugs, particularly if anyone else tries to make changes later.

My 2¢: I'd move most of this, i.e. what you want run on another goroutine, to a second private method, and then your`Run` (or doRun helper) looks like
```
<setup code>
g := ...
g.GoCtx(func(ctx) error { return ingestLoop(ctx, kvCh, ...) } )
g.GoCtx(func(ctx) error { return convertLoop(ctx, input, kvCh, ...)} )
g.Wait()

```

**Review by adityamaru27 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38374)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27 and @dt)

___
*[pkg/sql/create_table.go, line 1652 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LinEm374iTp6ZatP2D3:-LjIcAkA2QWAV6-TozMV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c528870bdbc83caf174b904bcf17573a12f8bac/pkg/sql/create_table.go#L1652)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

for better or worse, most of the time we define processors in `distsqlrun` and only use the function-pointer injection route for the ones that have to be defined elsewhere due to hard rules about dependency edges, like OSS vs CCL. Is there anything motivating putting this in `sql` instead of `distsqlrun`? Anything it depends on in `sql`?

in either case, I'd pull it into own file.
</blockquote></details>

Done.
___
*[pkg/sql/distsqlrun/processors.go, line 1244 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LiFVZ5UAWbBVxQCZzQr:-LjIcCi1CiAaCQShpIFi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d38628e6617e855bb05efd53dc7d79133640f19/pkg/sql/distsqlrun/processors.go#L1244)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

This is usually done to get around CCL vs OSS, where we inject the proc via this hook and an init func. This is a pure OSS proc, so I don't think we need that song and dance. 
</blockquote></details>

Done.
___
*[pkg/sql/bulk_row_writer.go, line 70 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LjHVsCs6cgH4uHDhuKZ:-LjIHTHVBtrCeX1e-KlZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c50c53cd89c44fcfe4b2c2cdef6716b23a0a277/pkg/sql/bulk_row_writer.go#L70)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: I'd call this method `ingestLoop` or something descriptive since it doesn't really _run_ the whole process.

That said, I think you're also on the right track going with the pattern of a `doRun` method that has an error return and can thus have early returns on error, since you can then just call it in `Run` and cleanup some of the named errors and conditionals deciding what to return.
</blockquote></details>

Done.
___
*[pkg/sql/bulk_row_writer.go, line 118 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38374#-LjHXATW-gBbp6EwKhwz:-LjIHVgEC1_NFEE2s4nI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c50c53cd89c44fcfe4b2c2cdef6716b23a0a277/pkg/sql/bulk_row_writer.go#L118)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

This is a lot of inline anon function along with the added layer below for the GoCtx call.

In my experience, big inline funcs, particularly closing over each other, can make it tricky to see what vars closed over vs local, what is shadowed, etc. and can sometimes lead to subtle bugs, particularly if anyone else tries to make changes later.

My 2¢: I'd move most of this, i.e. what you want run on another goroutine, to a second private method, and then your`Run` (or doRun helper) looks like
```
<setup code>
g := ...
g.GoCtx(func(ctx) error { return ingestLoop(ctx, kvCh, ...) } )
g.GoCtx(func(ctx) error { return convertLoop(ctx, input, kvCh, ...)} )
g.Wait()

```
</blockquote></details>

Done.

**Comment by adityamaru27** [comment]
TFTR
bors=r+

**Pull request merged**